### PR TITLE
Add symbiflow-vtr-optimized with PGO enabled

### PR DESCRIPTION
**Pull request by HackerFoo**
This builds VPR, and then collects profile data, and uses that to compile VPR again with profile-guided optimization.

### CODE REVIEW

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [inline symbiflow-vtr-optimized/build.sh]: This workaround could be used for other packages until https://github.com/conda/conda-build/issues/4118 is resolved.

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [inline symbiflow-vtr-optimized/build.sh:43]: This task runs in parallel with the first build, so that downloading and unpacking the SymbiFlow install package doesn't take any additional time.

**Review by litghost — CHANGES_REQUESTED** [review]
The eblif should be either sourced from another repo, or compressed.

File: symbiflow-vtr-optimized/build.sh
Comment (litghost) [reply symbiflow-vtr-optimized/build.sh]: Why do we need to patch this at all?  It has been merged upstream?

File: symbiflow-vtr-optimized/condarc
Comment (litghost) [inline symbiflow-vtr-optimized/condarc:2]: What are you pulling in from `conda-forge`?

File: symbiflow-vtr-optimized/meta.yaml
Comment (litghost) [inline symbiflow-vtr-optimized/meta.yaml]: Why are you using a git hash instead of a branch name?

**Review by litghost — CHANGES_REQUESTED** [review]
Comments in

File: symbiflow-vtr-optimized/meta.yaml
Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: This is the same hash used in symbiflow-arch-defs:
https://github.com/SymbiFlow/symbiflow-arch-defs/blob/991c74f0e1d579f143878b22d72ab65235f3c18c/environment.yml#L9

`master+wip` doesn't work. Would you like me to create a `master+wip-stable` branch?

Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: Why doesn't `master+wip` work?  @acomodi PR here was green: https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1711

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply symbiflow-vtr-optimized/build.sh]: It hasn't been merged into SymbiFlow/vtr-verilog-to-routing yet, though. This is temporary to avoid blocking on that.

File: symbiflow-vtr-optimized/condarc
Comment (HackerFoo) [reply symbiflow-vtr-optimized/condarc:2]: It's for `python-constraint`: https://travis-ci.com/github/SymbiFlow/conda-packages/jobs/441871052

File: symbiflow-vtr-optimized/meta.yaml
Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: It doesn't work in this test: https://travis-ci.com/github/SymbiFlow/conda-packages/jobs/441888559

Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: Here's a test with the architecture definition files from https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1711: https://travis-ci.com/github/SymbiFlow/conda-packages/jobs/441905520

Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: So fix it?  Or an absolute minimum, file a bug?  If you believe you have a reproducible error caused by https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1711 , that is something that needs to be resolved.

File: symbiflow-vtr-optimized/build.sh
Comment (litghost) [reply symbiflow-vtr-optimized/build.sh]: Part of the work is getting these fixes merged on SymbiFlow/vtr-verilog-to-routing?  I haven't seen you open a PR on SymbiFlow/vtr-verilog-to-routing integrating the upstream changes, so I'm confused by what you are doing here.

File: symbiflow-vtr-optimized/meta.yaml
Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: > from https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1711: https://travis-ci.com/github/SymbiFlow/conda-packages/jobs/441905520

This test failure is a result of using a eblif generated from an earlier version of arch-defs.  This highlights the fragility of using a frozen eblif.  What is your procedure for regenerating the eblif when arch-defs is updated?

Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: I think there is some confusion. The bug is in SymbiFlow/vtr-verilog-to-routing master+wip, which I think is indicated by the SymbiFlow+VtR tests that we know to be failing.

This uses the same version of [SymbiFlow/vtr-verilog-to-routing](https://github.com/SymbiFlow/vtr-verilog-to-routing/commit/f1a3bcc2a8aa93d88cb26ca7f1853962fff57c32) that the latest (including #1711) symbiflow-arch-defs uses.

Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: So the link you gave for the 1711 CI result failure was likely caused because the eblif was out of sync, specifically:

```
Type: Blif file
File: top.eblif
Line: 59289
Message: Found no matching port 'CO_CHAIN' on architecture model 'CARRY4_VPR'
```

Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: The reason this hasn't come up before is that building the Conda package does not run any tests with an xc7 architecture; adding these tests would likely cause the other symbiflow-vtr-* packages to fail as well.

Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: And for reference, `symbiflow+vtr` from #1711 was green: https://source.cloud.google.com/results/invocations/bc061b4a-3fb7-447b-9c3f-8237ecbed644/targets

Comment (litghost) [reply symbiflow-vtr-optimized/meta.yaml]: > The reason this hasn't come up before is that building the Conda package does not run any tests with an xc7 architecture; adding these tests would likely cause the other symbiflow-vtr-* packages to fail as well.

Totally agree with the above statement.  So what is your answer to solving this problem?

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply symbiflow-vtr-optimized/build.sh]: I just submitted a PR: https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/558

But it won't work even after that is merged, because master+wip is broken.

Comment (litghost) [reply symbiflow-vtr-optimized/build.sh]: > because master+wip is broken.

Is it though?  I just linked you to a CI showing it wasn't.  If you believe differently, please fix the issue or fill an issue showing the problem.

File: symbiflow-vtr-optimized/meta.yaml
Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: I will add a test to build.sh once this is in: https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1564

Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: Yes, I need to get a new eblif from the latest symbiflow-arch-defs.

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply symbiflow-vtr-optimized/build.sh]: I think I need a new eblif now that VPR has been fixed.

File: symbiflow-vtr-optimized/meta.yaml
Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: I've tried an eblif built from https://github.com/SymbiFlow/symbiflow-arch-defs/commit/d299ba7792fc4d3ccd9113b1629ccb21d877e7dd with https://github.com/SymbiFlow/vtr-verilog-to-routing/commit/1243707ceacab205ca49c58c10d004590fdaf9b8, the same versions as used in https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1711 from the VtR+SymbiFlow build log, and it still won't build. I suspect a difference in the `symbiflow_*` scripts, so I may need to copy the exact commands used from the symbiflow-arch-defs build.

Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: I may be able to get fpga-tool-perf working now, so I'll use the eblif from that instead.

File: symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply symbiflow-vtr-optimized/build.sh]: I've updated the eblif and removed this patch.

File: symbiflow-vtr-optimized/meta.yaml
Comment (HackerFoo) [reply symbiflow-vtr-optimized/meta.yaml]: I now have this working with master+wip.

**Comment by HackerFoo** [comment]
> The eblif should be either sourced from another repo, or compressed.

The eblif is compressed now (180KB).

**Comment by HackerFoo** [comment]
The `symbiflow-vtr-optimized` package builds successfully with `conda build` on my machine.

**Comment by HackerFoo** [comment]
I've added an issue to update this after https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1564 is merged: https://github.com/SymbiFlow/conda-packages/issues/155

**Review by litghost — COMMENTED** [review]
Couple things. 

stdout is too long.  Consider piping to a file and greping for error/warnings, etc?  Or just redirect to a file and output something when an error occurs?

CI is red:

```
The job exceeded the maximum time limit for jobs, and has been terminated.
```
This may be because of the stdout length?

**Comment by mithro** [comment]
> ```
> The job exceeded the maximum time limit for jobs, and has been terminated.
> ```

GitHub Actions has significantly more powerful hardware and a much longer timeout, so I think we might want to wait until after that is landed so we can take advantage of that?

**Comment by HackerFoo** [comment]
This PR is superseded by https://github.com/litex-hub/litex-conda-eda/pull/60

**Pull request closed without merging**
